src/pathpyG/core/WalkData.py

```diff
@@ -0,0 +1,234 @@
+from __future__ import annotations
+from typing import (
+    TYPE_CHECKING,
+    Dict,
+    List,
+    Tuple,
+    Union,
+    Any,
+    Optional,
+    Generator,
+)
+
+import torch
+from torch import IntTensor, Tensor, cat
+from torch_geometric import EdgeIndex
+
+from pathpyG.utils.config import config
+from pathpyG.core.IndexMap import IndexMap
+from pathpyG.core.PathData import PathData
+
+class WalkData(PathData):
+
+    def add(self, p: Tensor, freq: int = 1) -> None:
+        """Add an observation of a path or a walk in a graph based on a tensor representation
+
+        This method adds an observation of a walk, i.e. a sequence of not necessarily
+        unique nodes traversed in a graph. A walk of length l is represented as ordered
+        edge index tensor of size (2,l) where l is the number of traversed edges.
+        Walks can be associated with an integer that captures the observation frequency.
+
+        Since walks are a generalization of paths that allows for multiple traversals of
+        nodes, walks can be naturally used to store paths in a graph.
+
+        Walks can be seen as a special case of DAGs where the in- and out-degree of all
+        nodes is one. However, for a walk a higher-order model can be computed much more
+        efficiently using a GPU-based 1D convolution operation. It is thus advisable to
+        represent path data as walks whenever possible.
+
+        Args:
+            p:  An ordered edge index with size (2,l) that represents the sequence
+                in which a walk or path traverses the nodes of a graph.
+            freq:   The number of times this walk has been observed.
+
+        Example:
+            Assuming a `node_id` mapping of `['A', 'B', 'C', 'D']` the following snippet
+            stores three observations of the walk `A` --> `C` --> `D`:
+                ```py
+                import pathpyG as pp
+
+                paths = pp.PathData()
+                paths.add_walk(torch.tensor([[0, 2],[2, 3]]), freq=5)
```

Comment:
```suggestion
            Assuming a `node_id` mapping of `['A', 'B', 'C', 'D']` the following snippet
            stores four observations of the walk `A` --> `C` --> `D` --> `A`:
                ```py
                import pathpyG as pp

                paths = pp.PathData()
                paths.add_walk(torch.tensor([[0, 2, 3],[2, 3, 0]]), freq=5)
```
This way, the dimensions are not (2,2) which would make it very easy to confuse them. At least for me 😅